src/cli/utils.ts

```diff
@@ -0,0 +1,186 @@
+import {
+  BallotType,
+  CandidateContest,
+  Election,
+  getContests,
+  getBallotStyle,
+  getPrecinctById,
+  Dictionary,
+} from '../election'
+
+import { randomBase64 } from '../utils/random'
+
+export interface CastVoteRecordLocale {
+  primary: string
+  secondary?: string
+}
+
+export interface CastVoteRecord
+  extends Dictionary<
+    string | string[] | boolean | number | CastVoteRecordLocale
+  > {
+  _precinctId: string
+  _ballotStyleId: string
+  _ballotId: string
+  _testBallot: boolean
+  _scannerId: string
+  _pageNumber?: number
+  _locale?: CastVoteRecordLocale
+}
+
+interface objectWithSortkey<T> {
+  option: T
+  sortkey: number
+}
+
+export function trueWithPercentageProbability(percentage: number): boolean {
+  return Math.random() < percentage / 100
+}
+
+export function randomIntegerLessThan(max: number): number {
+  return Math.floor(Math.random() * max)
+}
+
+export function pickAtRandom<T>(options: T[], num: number): T[] {
+  if (num === 0) {
+    return []
+  }
+
+  const optionsWithRandomness = options.map((o) => {
+    return { option: o, sortkey: Math.random() } as objectWithSortkey<T>
+  })
+  optionsWithRandomness.sort((first, second) =>
+    first.sortkey < second.sortkey ? -1 : 1
+  )
+  return optionsWithRandomness.map((or) => or.option).slice(0, num)
+}
+
+interface GenerateCVRParams {
+  election: Election
+  precinctId: string
+  ballotStyleId: string
+  hasOvervote?: boolean
+  hasUndervote?: boolean
+  hasWriteins?: boolean
+  fromMultipage?: boolean
+  isTestBallot?: boolean
+  ballotType?: BallotType
+}
+
+export const generateCVR = ({
+  election,
+  precinctId,
+  ballotStyleId,
+  isTestBallot = true,
+  hasOvervote = false,
+  hasUndervote = false,
+  hasWriteins = false,
+  fromMultipage = false,
+}: // TODO: include ballot type?
+GenerateCVRParams): CastVoteRecord | undefined => {
+  const ballotStyle = getBallotStyle({ election, ballotStyleId })
+  const precinct = getPrecinctById({ election, precinctId })
+
+  if (!ballotStyle || !precinct) {
+    return
+  }
+
+  const votes: Dictionary<string[]> = {}
+
+  // get all the contests
+  const contests = getContests({ ballotStyle, election })
+
+  if (contests.length === 0) return
+
+  const overvoteContestId = hasOvervote
+    ? pickAtRandom(contests, 1)[0].id
+    : undefined
+
+  // let's not do too many undervotes, 1/3 at most
+  const numUndervotes = hasUndervote
+    ? randomIntegerLessThan(contests.length / 3)
+    : 0
+  const undervoteContestIds = hasUndervote
+    ? pickAtRandom(contests, numUndervotes).map((c) => c.id)
+    : []
+
+  // let's do even fewer write-ins, 1/5 at most
+  const contestsWithWriteins = contests.filter(
+    (c) => c.type === 'candidate' && (c as CandidateContest).allowWriteIns
```

Comment:
```suggestion
    (c) => c.type === 'candidate' && c.allowWriteIns
```

Cast isn't necessary.